modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessor.java

```diff
@@ -1051,21 +1060,151 @@ private void validateNameFields(TypeElement clazz, List<VariableElement> fields)
+                        clazz.getQualifiedName(),
+                        superClazz.getQualifiedName()
+                ));
+            }
+
+            if (!collectFieldsWithAnnotation(superClazzFields, InternalId.class).isEmpty()) {
+                throw new ConfigurationProcessorException(String.format(
+                        invalidFieldInSuperClassFormat,
+                        simpleName(InternalId.class),
+                        clazz.getQualifiedName(),
+                        superClazz.getQualifiedName()
+                ));
+            }
+        }
+    }
+
+    /**
+     * Checks configuration schema with {@link Config}.
+     *
+     * @param clazz Type element under validation.
+     * @param fields Non-static fields of the class under validation.
+     * @throws ConfigurationProcessorException If validation fails.
+     */
+    private void validateConfig(TypeElement clazz, List<VariableElement> fields) throws ConfigurationProcessorException {
+        checkIncompatibleClassAnnotations(
+                clazz,
+                Config.class,
+                incompatibleSchemaClassAnnotations(Config.class)
+        );
+
+        checkNotContainsPolymorphicIdField(clazz, Config.class, fields);
+
+        TypeElement superClazz = superClass(clazz);
+
+        if (!isClass(superClazz.asType(), Object.class)) {
+            checkSuperclassContainAnyAnnotation(clazz, superClazz, AbstractConfiguration.class);
+
+            List<VariableElement> superClazzFields = fields(superClazz);
+
+            checkNoConflictFieldNames(clazz, superClazz, fields, superClazzFields);
+
+            String fieldAlreadyPresentInSuperClassFormat = "Field with %s is already present in the superclass [class=%s, superClass=%s]";
+
+            if (!collectFieldsWithAnnotation(superClazzFields, InjectedName.class).isEmpty()
+                    && !collectFieldsWithAnnotation(fields, InjectedName.class).isEmpty()) {
+                throw new ConfigurationProcessorException(String.format(
+                        fieldAlreadyPresentInSuperClassFormat,
+                        simpleName(InjectedName.class),
+                        clazz.getQualifiedName(),
+                        superClazz.getQualifiedName()
+                ));
+            }
+
+            if (!collectFieldsWithAnnotation(superClazzFields, InternalId.class).isEmpty()
+                    && !collectFieldsWithAnnotation(fields, InternalId.class).isEmpty()) {
+                throw new ConfigurationProcessorException(String.format(
+                        fieldAlreadyPresentInSuperClassFormat,
+                        simpleName(InternalId.class),
+                        clazz.getQualifiedName(),
+                        superClazz.getQualifiedName()
+                ));
+            }
+        }
+    }
+
+    private Class<? extends Annotation>[] incompatibleSchemaClassAnnotations(Class<? extends Annotation>... compatibleAnnotations) {
```

Comment:
We can, but then it will be necessary to do `.toArray(Class[]::new)` everywhere, this is a private method, I suggest leaving it that way, wdyt?

Reply:
This method is only called as a parameter of `checkIncompatibleClassAnnotations`, so I think it can be called from there instead. But yes, you will then have to call `toArray` anyway, so my original comment can be considered obsolete

Reply:
Ok